locator/tablets.hh

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright (C) 2023-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#pragma once
+
+#include "dht/token.hh"
+#include "utils/small_vector.hh"
+#include "locator/host_id.hh"
+#include "dht/i_partitioner_fwd.hh"
+#include "schema/schema_fwd.hh"
+#include "utils/chunked_vector.hh"
+#include "utils/hash.hh"
+
+#include <seastar/core/reactor.hh>
+#include <seastar/util/log.hh>
+
+#include <vector>
+
+namespace locator {
+
+extern seastar::logger tablet_logger;
+
+using token = dht::token;
+
+// Identifies tablet within the scope of a single tablet_map,
+// which has a scope of (table_id, token metadata version).
+// Different tablets of different tables can have the same tablet_id.
+// Different tablets in subsequent token metadata version can have the same tablet_id.
+// When splitting a tablet, one of the new tablets (in the new token metadata version)
+// will have the same tablet_id as the old one.
+using tablet_id = size_t;
+
+struct tablet_replica {
+    host_id host;
+    shard_id shard;
+
+    bool operator==(const tablet_replica&) const = default;
+};
+
+std::ostream& operator<<(std::ostream&, const tablet_replica&);
+
+using tablet_replica_set = utils::small_vector<tablet_replica, 3>;
+
+/// Stores information about a single tablet.
+struct tablet_info {
+    tablet_replica_set replicas;
```

Comment:
since the tablet must not be replicated on multiple shards on the same host, how about using a `std::unordered_map<host_id, shard_id>`?
Does the order of replicas matter?

Reply:
If using a vector I'd add some checks in non-release modes around uniqueness of the host_id on the insertion and/or retrieval paths.

Reply:
Such checks probably belong to the tablet allocator which produces these, and involve not only uniqueness of the host, but also racks and DCs. There's `endpoints_check()` in network_topology_strategy_test.cc. Maybe each replication strategy should provide its sanity checks? 

I think this can be done incrementally.

Reply:
agree